### PR TITLE
MINIFICPP-1537 - Log heartbeats on demand

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by martinzink — COMMENTED** [review]
Looks good to me, however I think we could also add some tests as well, maybe something along the lines of IntegrationBase where you set the heartbeat.reporter.classes property and then verify the logs with verifyLogLinePresenceInPollTime

File: conf/minifi.properties [hunk 1]
```diff
@@ -67,6 +67,7 @@ nifi.content.repository.class.name=DatabaseContentRepository
 ## define protocol parameters
 ## The default is CoAP, if that extension is built. 
 ## Alternatively, you may use RESTSender if http-curl is built
+#nifi.c2.agent.protocol.class=RESTSender
```
Comment (arpadboda) [inline conf/minifi.properties:70]: This doesn't seem to be related

**Review by arpadboda — APPROVED** [review]
Looks good mostly, one minor comment

File: conf/minifi.properties (hunk 1 above)
Comment (adamdebreceni) [reply conf/minifi.properties]: it became obsolete in light of #1058, removed

**Comment by adamdebreceni** [comment]
> Looks good to me, however I think we could also add some tests as well, maybe something along the lines of IntegrationBase where you set the heartbeat.reporter.classes property and then verify the logs with verifyLogLinePresenceInPollTime

**Review by martinzink — APPROVED** [review]

File: libminifi/src/c2/HeartbeatJsonSerializer.cpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+      return "start";
+    case Operation::UPDATE:
+      return "update";
+    case Operation::PAUSE:
+      return "pause";
+    case Operation::RESUME:
+      return "resume";
+    default:
+      return "heartbeat";
+  }
```
Comment (fgerlits) [inline libminifi/src/c2/HeartbeatJsonSerializer.cpp:52]: I know this is old code and you just moved it, but this function converts some enumeration values like CLEAR and TRANSFER into "heartbeat" -- is this OK?

I would expect the switch statement to
* cover all possible enum values
* list values in the same order as in the enum definition
* not have a default, or have a default like "UNKNOWN".

If this behavior is intentional, it's probably worth adding a comment to explain why.

File: libminifi/src/c2/protocols/RESTProtocol.cpp [hunk 3]
```diff
@@ -318,133 +178,7 @@ bool RESTProtocol::containsPayload(const C2Payload &o) {
...
-      return "pause";
-    case Operation::RESUME:
-      return "resume";
-    default:
-      return "heartbeat";
-  }
-}
-
-Operation RESTProtocol::stringToOperation(const std::string str) {
+Operation RESTProtocol::stringToOperation(const std::string& str) {
```
Comment (fgerlits) [inline libminifi/src/c2/protocols/RESTProtocol.cpp:181]: I think `stringToOperation()` should be in the same class as `getOperation()` [and cover the same values, in the same order].

File: extensions/http-curl/tests/C2LogHeartbeatTest.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  void runAssertions() override {
+    using org::apache::nifi::minifi::utils::verifyLogLinePresenceInPollTime;
+    assert(verifyLogLinePresenceInPollTime(
+        std::chrono::milliseconds(wait_time_),
+        "\"operation\": \"heartbeat\""));
+  }
+
+  void configureC2() override {
+    VerifyC2Base::configureC2();
+    configuration->set("c2.agent.heartbeat.reporter.classes", "HeartbeatLogger");
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/http-curl/tests/C2LogHeartbeatTest.cpp]: done

File: libminifi/src/c2/HeartbeatJsonSerializer.cpp (hunk 2 above)
Comment (adamdebreceni) [reply libminifi/src/c2/HeartbeatJsonSerializer.cpp]: unclear why "heartbeat" was used in those instances, those features are probably not tested, made Operation into a SMART_ENUM with explicit fallback for stringifying invalid values

File: libminifi/src/c2/protocols/RESTProtocol.cpp (hunk 3 above)
Comment (adamdebreceni) [reply libminifi/src/c2/protocols/RESTProtocol.cpp]: made Operation enum into a smart enum, so that string <-> enum conversions are centralized

File: libminifi/include/c2/HeartbeatJsonSerializer.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+#ifdef RAPIDJSON_ASSERT
+#undef RAPIDJSON_ASSERT
+#endif
+#define RAPIDJSON_ASSERT(x) if(!(x)) throw std::logic_error("rapidjson exception"); //NOLINT
+
+#ifdef RAPIDJSON_HAS_STDSTRING
+#undef RAPIDJSON_HAS_STDSTRING
+#endif
+#define RAPIDJSON_HAS_STDSTRING 1
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/c2/HeartbeatJsonSerializer.h:33]: unclear why it was not used before, enabling it simplifies many json operations

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/c2/HeartbeatJsonSerializer.h]: reverted the change, we could change it in a separate PR, updating every usage

File: libminifi/src/c2/protocols/RESTProtocol.cpp (hunk 3 above)
Comment (fgerlits) [reply libminifi/src/c2/protocols/RESTProtocol.cpp]: It looks like changes are needed in other classes to use the new enum correctly, as some CI jobs are failing.

**Review by fgerlits — APPROVED** [review]

File: libminifi/include/core/logging/Logger.h [hunk 6]
```diff
@@ -68,13 +69,32 @@ inline T conditional_conversion(T t) {
...
+      return std::string("Error while formatting log message");
+    }
+    std::vector<char> buffer(size_needed + 1);  // extra '\0' character
+    std::snprintf(buffer.data(), buffer.size(), format_str, conditional_conversion(std::forward<Args>(args))...);
+    return std::string(buffer.data());
+  }
+  // use dynamic but fix-sized buffer
+  std::vector<char> buffer(max_size);
+  std::snprintf(buffer.data(), buffer.size(), format_str, conditional_conversion(std::forward<Args>(args))...);
+  return std::string(buffer.data());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/logging/Logger.h]: changed them to use either the `(const char*, size_t)` or the `(It first, It last)` constructors

File: libminifi/include/core/logging/Logger.h [hunk 7]
```diff
@@ -68,13 +69,32 @@ inline T conditional_conversion(T t) {
...
+  if (0 <= max_size && max_size <= LOG_BUFFER_SIZE) {
+    // use static buffer
+    char buf[LOG_BUFFER_SIZE + 1];
+    std::snprintf(buf, max_size + 1, format_str, conditional_conversion(std::forward<Args>(args))...);
+    return std::string(buf);
+  }
+  // use dynamically allocated buffer
+  if (max_size < 0) {
+    // query what buffer size we need
+    int size_needed = std::snprintf(nullptr, 0, format_str, conditional_conversion(std::forward<Args>(args))...);
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/logging/Logger.h]: done

File: libminifi/src/Configure.cpp [hunk 8]
```diff
@@ -37,6 +37,10 @@ bool Configure::get(const std::string& key, std::string& value) const {
 
 bool Configure::get(const std::string& key, const std::string& alternate_key, std::string& value) const {
   if (get(key, value)) {
+    if (get(alternate_key)) {
+      const auto logger = logging::LoggerFactory<Configure>::getLogger();
+      logger->log_warn("Both the property '%s' and an alternate property '%s' are set. Using '%s'.", key, alternate_key, key);
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/Configure.cpp:42]: I was under the impression that these "non-nifi-prefixed" property names are deprecated and will be removed in the future, as the user can specify these "ignored" properties I feel like the fact that we are discarding them is worthy of their attention

File: libminifi/src/c2/HeartbeatJsonSerializer.cpp [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    case state::UpdateState::PARTIALLY_APPLIED:
+      state_str = "PARTIALLY_APPLIED";
+      break;
+    case state::UpdateState::READ_ERROR:
+      state_str = "OPERATION_NOT_UNDERSTOOD";
+      break;
+    case state::UpdateState::SET_ERROR:
+    default:
+      state_str = "NOT_APPLIED";
+  }
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/c2/HeartbeatJsonSerializer.cpp]: done

File: libminifi/include/utils/Enum.h [hunk 10]
```diff
@@ -60,6 +63,22 @@ namespace utils {
...
+  fn(_5) delim() fn(_6) delim() fn(_7) delim() fn(_8) delim() \
+  fn(_9) delim() fn(_10)
+#define FOR_EACH_11(fn, delim, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11) \
+  fn(_1) delim() fn(_2) delim() fn(_3) delim() fn(_4) delim() \
+  fn(_5) delim() fn(_6) delim() fn(_7) delim() fn(_8) delim() \
+  fn(_9) delim() fn(_10) delim() fn(_11)
+#define FOR_EACH_12(fn, delim, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12) \
+  fn(_1) delim() fn(_2) delim() fn(_3) delim() fn(_4) delim() \
+  fn(_5) delim() fn(_6) delim() fn(_7) delim() fn(_8) delim() \
+  fn(_9) delim() fn(_10) delim() fn(_11) delim() fn(_12)
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/utils/Enum.h]: I cannot recall what was the reasoning behind the current non-"recursive" macros, changed them

File: libminifi/src/Configure.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/Configure.cpp:42]: Is this function only used in those cases? My impression was just a general function with a fallback key option, as the signature doesn't imply `alternate_key` being deprecated.

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/Configure.cpp:42]: see `C2.md`:
> Release 0.6.0: Please note that all c2 properties now exist as `nifi.c2.*`. If your configuration properties files contain the former naming convention of `c2.*`, we will continue to support that as an alternate key, but you are encouraged to switch your configuration options as soon as possible.

although it does not say "deprecated" per se

File: libminifi/include/core/logging/Logger.h [hunk 11]
```diff
@@ -68,13 +69,32 @@ inline T conditional_conversion(T t) {
...
+    return std::string(buf, result);
+  }
+  if (max_size >= 0 && max_size <= LOG_BUFFER_SIZE) {
+    // static buffer was already larger than allowed, use the filled buffer
+    return std::string(buf, LOG_BUFFER_SIZE);
+  }
+  // try to use dynamic buffer
+  size_t dynamic_buffer_size = max_size < 0 ? result : std::min(result, max_size);
+  std::vector<char> buffer(dynamic_buffer_size + 1);  // extra '\0' character
+  result = std::snprintf(buffer.data(), buffer.size(), format_str, conditional_conversion(std::forward<Args>(args))...);
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/logging/Logger.h:90]: this thought crossed my mind, but this explicitly says that modifying through the const overload is undefined behavior, I would wait until C++17

File: libminifi/src/Configure.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/Configure.cpp:42]: I checked that the function is only used for this purpose, so now I think it's good.

**Review by szaszm — APPROVED** [review]
